chat.js

```diff
@@ -948,3 +992,39 @@ Chat.toDurationString = function (number, options) {
 	}
 	return parts.slice(positiveIndex).reverse().map((value, index) => value ? value + " " + unitNames[index] + (value > 1 ? "s" : "") : "").reverse().slice(0, precision).join(" ").trim();
 };
+
+/**
+ * @param  {string} str
+ * @return {string}
+ */
+Chat.md5 = function (str) {
+	 let generator = crypto.createHash('md5');
+	 generator.update(str);
+	 return generator.digest('hex');
+};
+
+/**
+ * @param {Room} room
+ * @param {string} checksum
+ */
+Chat.r9kLoadPermanent = function (room) {
+	let data;
+	try {
+		data = fs.readFileSync(path.resolve(R9K_DIRECTORY, `${room.id}.txt`));
+	} catch (e) {
+		return;
+	}
+	data = String(data).split("\n").map(val => val.trim());
+	if (data.some(val => val.length !== 16)) return;
+	room.r9kPermanentCache = new Set(data);
+};
+
+/**
+ * @param {Room} room
+ * @param {string} checksum
+ */
+Chat.r9kAppendPermanent = function (room, checksum) {
+	room.r9kPermanentCache.add(checksum);
+
+	fs.appendFile(path.resolve(R9K_DIRECTORY, `${room.id}.txt`), `${checksum}\n`, () => {});
+};
```

Comment:
It's because GitHub is buggy and doesn't display the "you forgot the trailing newline" symbol in the overview. In the code review page, it looks like this:

![image](https://cloud.githubusercontent.com/assets/551184/21753319/b6ea0b82-d59f-11e6-8c4c-9f9c5a8d603a.png)

Which makes the issue pretty obvious.